### PR TITLE
Adding replication ability to valkey test case, changing waiter functionality for more flexibility in the future

### PR DESCRIPTION
**Pull request by zackcam**
Added a class for replication in valkey_test_case.py as well as a simple test that checks it works as intended. In addition to this changed how waiters worked to allow more cases to be added in the future with different operators. 

Due to the change in waiters the timing test in test_save_and_restore was now quicker causing it to fail so changed the position of the wait in order for the test to work again.

### CODE REVIEW

File: tests/test_replication.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        bf_exists_result = self.client.execute_command('BF.EXISTS key item1')
+        bf_non_added_exists_result = self.client.execute_command('BF.EXISTS key item2')
+
+        self.waitForReplicaToSyncUp(self.replicas[0])
+
+        bf_replica_exists_result = self.replicas[0].client.execute_command('BF.EXISTS key item1')
+        assert bf_exists_result == bf_replica_exists_result
+        bf_replica_non_added_exists_result = self.replicas[0].client.execute_command('BF.EXISTS key item2')
+        assert bf_non_added_exists_result == bf_replica_non_added_exists_result
+
```
Comment (YueTang-Vanessa) [inline tests/test_replication.py]: Can we add an assertion for `BF.INFO key` results?

File: tests/valkeytests/valkey_test_case.py [hunk 2]
```diff
@@ -339,10 +367,16 @@ def wait_for_key(self, key, value):
...
         c = ValkeyClient.create_from_server(self)
-        if not self._waitForPing(c):
-            raise RuntimeError("Failed to connect or ping server")
+        try:
+            self._waitForPing(c)
+        except WaitTimeout:
+             raise RuntimeError("Failed to connect or ping server")
         self.client = c
 
+    def wait_for_all_replicas_online(self, n):
```
Comment (YueTang-Vanessa) [inline tests/valkeytests/valkey_test_case.py]: Same as previous, can you change the n to a more clear argument name?

File: tests/valkeytests/valkey_test_case.py [hunk 3]
```diff
@@ -254,7 +277,10 @@ def children_pids(self):
                 children.append(line)
         return children
 
-    @wait(1, 60) # wait upto 30 sec checking every sec
+    def wait_for_replicas(self, n):
```
Comment (YueTang-Vanessa) [inline tests/valkeytests/valkey_test_case.py]: Can you change the `n` to a more clear argument name? Like `num_of_nodes` or `node_count` etc.

File: tests/valkeytests/util/waiters.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    asserts that this happens within a timeout.
+    """
+    _wait_for(func, expected_value, **kwargs)
+
+def wait_for_true(func, **kwargs):
+    """
+    Wait for a given function to return True, and
+    asserts that this happens within a timeout.
+    """
+    _wait_for(func, **kwargs)
```
Comment (KarthikSubbarao) [inline tests/valkeytests/util/waiters.py]: Let's add a new line after this

File: tests/test_replication.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import os
+
+class TestBloomReplication(ReplicationTestCase):
+
+    def get_custom_args(self):
+        self.set_server_version(os.environ['SERVER_VERSION'])
+        return {
+            'loadmodule': os.getenv('MODULE_PATH'),
+        }
+    def test_replication_success(self):
```
Comment (KarthikSubbarao) [inline tests/test_replication.py:14]: Let's add a new line

File: tests/test_replication.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        bf_add_result = self.client.execute_command('BF.ADD key item1')
+        assert bf_add_result == 1
```
Comment (KarthikSubbarao) [inline tests/test_replication.py:16]: We can remove this line since a false positive could occur here

File: tests/test_save_and_restore.py [hunk 7]
```diff
@@ -24,8 +24,8 @@ def test_basic_save_and_restore(self):
         # save rdb, restart sever
         client.bgsave()
         self.server.wait_for_save_done()
+        time.sleep(1)
```
Comment (KarthikSubbarao) [inline tests/test_save_and_restore.py:28]: Is this needed?

File: tests/valkeytests/valkey_test_case.py [hunk 8]
```diff
@@ -482,3 +568,122 @@ def teardown(self):
...
+        default_masterhost = None
+        default_port = None
+        if connection_type == 'tcp':
+            if hasattr(self.server, 'bind_ip'):
+                default_masterhost = self.server.bind_ip
+            if hasattr(self.server, 'port'):
+                default_port = self.server.port
+        elif connection_type == 'unix':
+            default_masterhost = self.server.args["unixsocket"]
+            default_port = 0    # Redis treats the hostname as a unix socket path if the port is zero.
```
Comment (KarthikSubbarao) [inline tests/valkeytests/valkey_test_case.py]: "Valkey treats...."

**Pull request merged**